### ISSUE TITLE
PEP 487

### ISSUE THREAD

**Issue by NeilGirdhar** [ipython/traitlets#258]
This is not really an issue, but I wasn't sure where to post it.  With [PEP 487](https://www.python.org/dev/peps/pep-0487/) accepted into Python 3.6, it might be worth thinking about how traitlets can benefit if at all.

**Comment by rmorshea** [comment]
Since traitlets already has a framework for descriptor setup and still suport Python<3.6, we are probably limited to utilizing `otherargs` in `type.__new__` (for what I'm not sure). However something noteworthy is that PEP 487 implements a subclass setup phase, while traitlets only implements an "owner class" setup phase in `class_init`.

Being able to distinguish between an owner class, and any other subclass remains significant, for traitlets because there's a need for inheritance enforcement, but it does suggest that it subclass setup may be a useful feature. In other words we may want a `subclass_init` method for the same purposes it serves in PEP 487.

**Comment by NeilGirdhar** [comment]
@rmorshea Thanks for sharing your thoughts.  Feel free to close this since this isn't a real issue :)

**Comment by rmorshea** [comment]
@NeilGirdhar though I contribute a lot here, I actually don't have collaborator access.

I do think this brings up an interesting point about including a `subclass_init` method though.

**Comment by SylvainCorlay** [comment]
@NeilGirdhar is the state of the PEP definitive?

**Comment by NeilGirdhar** [comment]
@SylvainCorlay I don't know, but the very recent discussion is here: https://mail.python.org/pipermail/python-dev/2016-July/thread.html#145466.

**Comment by NeilGirdhar** [comment]
@SylvainCorlay When the PEP is totally final, it might be worth renaming the methods in traitlets to match the corresponding methods in Python 3.6?  And then maybe use a more lightweight version of MetaHasDescriptors for 3.6?  I wonder if this would be possible.

**Comment by SylvainCorlay** [comment]
Yeah, I would really like to not use our custom meta class at all in Python 3.6.

**Comment by SylvainCorlay** [comment]
(although traitlets is at the very bottom of the entire Jupyter stack so we want to be very careful with it).

**Comment by NeilGirdhar** [comment]
@SylvainCorlay Awesome.  I agree that it would be more elegant if you won't need your custom meta class.

**Comment by jasongrout** [comment]
@SylvainCorlay, we won't be able to switch to it any time soon. Unless a metaclass backports the features (is it possible? Anyone planning to do it?)

**Comment by SylvainCorlay** [comment]
We could have the `MetaHasDescriptors` converge towards Python 3.6' s `type` implementation.

**Comment by NeilGirdhar** [comment]
So, should `instance_init` be renamed `__set_name__` and not explicitly called in Python 3.6+ (since it is called automatically on all descriptors)?

**Comment by rmorshea** [comment]
Rather, `class_init` would be renamed `__set_name__` and `subclass_init` would be renamed `__init_subclass__`. To keep with the theme, we might choose `init_instance` instead of `instance_init`.

**Comment by NeilGirdhar** [comment]
Wait a second, `__set_name__` gets the object and the name.  `class_init` gets the class and the name, and `instantce_init` gets the object.  It seems like there's a different organization?

**Comment by rmorshea** [comment]
According to the PEP:

> upon class creation, a `__set_name__` hook is called

So it can't possible be the `obj` from `__get__(self, obj, cls)`.

**Comment by rmorshea** [comment]
@NeilGirdhar, I'm not sure if this is what you had meant, but `__init_subclass__` does not do what I thought. `__set_name__` is analogous to `class_init`, but `__init_subclass__` is **not** analogous to `subclass_init`.

**Comment by NeilGirdhar** [comment]
This is stale
